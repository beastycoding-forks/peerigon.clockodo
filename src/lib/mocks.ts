import { faker } from "@faker-js/faker";

export const ONE_DAY = 24 * 60 * 60 * 1000;
const MAX_ITERATION_COUNT = 10000;

export const isoDateFromDateTime = (dateTime: Date) => {
  return dateTime.toISOString().replace(/T.*/, "");
};

export const isoMonthFromDateTime = (dateTime: Date) => {
  return dateTime.toISOString().replace(/-\d\dT.*/, "");
};

export const isoDateFromTimestamp = (timestamp: number) => {
  return isoDateFromDateTime(new Date(timestamp));
};

export const timestampFromIsoDate = (isoDate: string) => {
  return new Date(isoDate).getTime();
};

export const endOfMonth = (date: Date) => {
  let startOfNextMonth = new Date(date.getTime());

  startOfNextMonth.setMonth(date.getMonth() + 1);

  if (startOfNextMonth.getMonth() === 0) {
    startOfNextMonth.setFullYear(date.getFullYear() + 1);
  }

  startOfNextMonth = new Date(isoMonthFromDateTime(startOfNextMonth));

  const endOfMonth = new Date(startOfNextMonth);

  endOfMonth.setTime(endOfMonth.getTime() - ONE_DAY / 2);

  return new Date(isoDateFromDateTime(endOfMonth));
};

export const endOfYear = (dateTime: Date) => {
  return new Date(`${dateTime.getFullYear()}-12-31T23:59:59.999Z`);
};

export const nextDay = (date: Date) => {
  return new Date(isoDateFromDateTime(new Date(date.getTime() + ONE_DAY)));
};

export const generateUnique = <Value>({
  count,
  generate,
}: {
  count: number;
  generate: () => Value;
}) => {
  const generated = new Set<Value>();

  for (let i = 0; generated.size < count; i++) {
    if (i === MAX_ITERATION_COUNT) {
      throw new Error(
        `Couldn't generate enough unique values before reaching the max iteration count.
This usually happens when the generated values create too many conflicts (e.g. too many dates on a small date range).`
      );
    }
    generated.add(generate());
  }

  return generated;
};

export const generateDayRange = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomDay = isoDateFromDateTime(faker.date.between(from, to));

        return timestampFromIsoDate(randomDay);
      },
    })
  ).sort();
};

export const generateMonthRange = ({
  count,
  between: [from, to],
}: {
  count: number;
  between: [Date, Date];
}) => {
  return Array.from(
    generateUnique({
      count,
      generate: () => {
        const randomMonth = isoMonthFromDateTime(faker.date.between(from, to));

        return timestampFromIsoDate(randomMonth);
      },
    })
  ).sort();
};

export const toPairs = <Item>(array: Array<Item>) => {
  if (array.length % 2 !== 0) {
    throw new Error("Cannot create pairs: Array length must be divisible by 2");
  }

  const count = array.length / 2;
  const pairs = new Array<[Item, Item]>(count);

  for (let i = 0; i < count; i++) {
    pairs[i] = [array[i * 2], array[i * 2 + 1]];
  }

  return pairs;
};
