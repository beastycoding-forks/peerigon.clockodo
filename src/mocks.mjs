// eslint-disable-next-line import/no-unresolved
export * from "./mocks.js";
