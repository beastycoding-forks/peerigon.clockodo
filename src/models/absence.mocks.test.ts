import { createAbsencesMocks } from "./absence.mocks.js";

describe("createAbsencesMocks()", () => {
  test("It creates realistic mocks", () => {
    expect(createAbsencesMocks({ count: 20 })).toMatchSnapshot();
  });
});
