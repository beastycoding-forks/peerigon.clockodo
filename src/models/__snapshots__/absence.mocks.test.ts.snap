// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAbsenceMocks() It creates realistic mocks 1`] = `
Array [
  Object {
    "approvedBy": 1,
    "countDays": 1,
    "countHours": null,
    "dateApproved": "2019-09-05",
    "dateEnquired": null,
    "dateSince": "2020-01-28",
    "dateUntil": "2020-01-31",
    "id": 0,
    "note": null,
    "status": 1,
    "type": 1,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-02-24",
    "dateUntil": "2020-02-24",
    "id": 1,
    "note": null,
    "status": 3,
    "type": 2,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-03-09",
    "dateUntil": "2020-03-09",
    "id": 2,
    "note": null,
    "status": 0,
    "type": 6,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 1,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-04-12",
    "dateUntil": "2020-04-14",
    "id": 3,
    "note": null,
    "status": 2,
    "type": 7,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "dateApproved": null,
    "dateEnquired": "2019-11-08",
    "dateSince": "2020-05-01",
    "dateUntil": "2020-05-01",
    "id": 4,
    "note": null,
    "sickNote": true,
    "status": 3,
    "type": 4,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 1,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-05-24",
    "dateUntil": "2020-05-25",
    "id": 5,
    "note": null,
    "status": 3,
    "type": 9,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": "2020-05-22",
    "dateEnquired": null,
    "dateSince": "2020-05-27",
    "dateUntil": "2020-05-27",
    "id": 6,
    "note": null,
    "status": 1,
    "type": 10,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 1,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-06-04",
    "dateUntil": "2020-06-05",
    "id": 7,
    "note": null,
    "status": 2,
    "type": 11,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-06-07",
    "dateUntil": "2020-06-07",
    "id": 8,
    "note": null,
    "status": 0,
    "type": 12,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-06-12",
    "dateUntil": "2020-06-12",
    "id": 9,
    "note": "totam officiis",
    "status": 2,
    "type": 13,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 13,
    "countHours": null,
    "dateApproved": "2020-03-05",
    "dateEnquired": null,
    "dateSince": "2020-06-28",
    "dateUntil": "2020-07-12",
    "id": 10,
    "note": null,
    "status": 1,
    "type": 14,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 4.479627713258378,
    "dateApproved": "2020-04-07",
    "dateEnquired": "2020-04-03",
    "dateSince": "2020-07-17",
    "dateUntil": "2020-07-20",
    "id": 11,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": null,
    "dateSince": "2020-08-04",
    "dateUntil": "2020-08-04",
    "id": 12,
    "note": null,
    "status": 3,
    "type": 1,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 7.963332051970065,
    "dateApproved": "2020-04-29",
    "dateEnquired": "2020-04-27",
    "dateSince": "2020-08-17",
    "dateUntil": "2020-08-19",
    "id": 13,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 7.012903982773423,
    "dateApproved": "2020-05-13",
    "dateEnquired": null,
    "dateSince": "2020-09-08",
    "dateUntil": "2020-09-09",
    "id": 14,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 4.153712440351955,
    "dateApproved": "2020-06-30",
    "dateEnquired": null,
    "dateSince": "2020-09-11",
    "dateUntil": "2020-09-11",
    "id": 15,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": null,
    "countHours": 2.643646778725088,
    "dateApproved": "2020-07-31",
    "dateEnquired": null,
    "dateSince": "2020-09-17",
    "dateUntil": "2020-09-17",
    "id": 16,
    "note": null,
    "status": 1,
    "type": 3,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 0.5,
    "countHours": null,
    "dateApproved": "2020-06-03",
    "dateEnquired": "2020-05-30",
    "dateSince": "2020-09-20",
    "dateUntil": "2020-09-20",
    "id": 17,
    "note": null,
    "status": 1,
    "type": 9,
    "usersId": 0,
  },
  Object {
    "approvedBy": null,
    "countDays": 1,
    "countHours": null,
    "dateApproved": null,
    "dateEnquired": "2020-07-02",
    "dateSince": "2020-09-26",
    "dateUntil": "2020-09-27",
    "id": 18,
    "note": null,
    "status": 0,
    "type": 10,
    "usersId": 0,
  },
  Object {
    "approvedBy": 1,
    "countDays": 15,
    "countHours": null,
    "dateApproved": "2020-11-22",
    "dateEnquired": "2020-11-18",
    "dateSince": "2020-11-23",
    "dateUntil": "2020-12-10",
    "id": 19,
    "note": null,
    "status": 1,
    "type": 11,
    "usersId": 0,
  },
]
`;
